service/controller/azuremachinepool/handler/migration/resource.go

```diff
@@ -0,0 +1,159 @@
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if !areReferencesUpdated(azureMachinePool) {
+		// Migration from old to new MachinePool is not completed. Cancel
+		// remaining reconciliation.
+		reconciliationcanceledcontext.SetCanceled(ctx)
+		return nil
+	}
+
+	// Now the new AzureMachinePool has been created with Spec copied from old
+	// AzureMachinePool, so here we want to:
+	// 1. Clone Status from old AzureMachinePool
+	// 2. Delete old AzureMachinePool
+
+	// Get old AzureMachinePool
+	namespacedName := types.NamespacedName{
+		Namespace: azureMachinePool.Namespace,
+		Name:      azureMachinePool.Name,
+	}
+	oldAzureMachinePool := &oldcapzexpv1alpha3.AzureMachinePool{}
+	err = r.ctrlClient.Get(ctx, namespacedName, oldAzureMachinePool)
+	if apierrors.IsNotFound(err) {
+		// Old AzureMachinePool not found, so nothing to do here.
+		return nil
+	} else if err != nil {
+		// Migration from old to new AzureMachinePool is not completed
+		// because we still didn't update the status in the new AzureMachinePool,
+		// so we cancel remaining reconciliation.
+		reconciliationcanceledcontext.SetCanceled(ctx)
+		return microerror.Mask(err)
+	}
+
+	// 1. Update status in new AzureMachinePool
+	if isStatusEmpty(azureMachinePool) {
+		err = cloneObject(&oldAzureMachinePool.Status, &azureMachinePool.Status)
+		if err != nil {
+			// Migration from old to new AzureMachinePool is not completed
+			// because we still didn't update the status in the new AzureMachinePool,
+			// so we cancel remaining reconciliation.
+			reconciliationcanceledcontext.SetCanceled(ctx)
+			return microerror.Mask(err)
+		}
+
+		// Update new AzureMachinePool
+		err = r.ctrlClient.Status().Update(ctx, &azureMachinePool)
+		if err != nil {
+			// Migration from old to new AzureMachinePool is not completed
+			// because we still didn't update the status in the new AzureMachinePool,
+			// so we cancel remaining reconciliation.
+			reconciliationcanceledcontext.SetCanceled(ctx)
+			return microerror.Mask(err)
+		}
+	}
+
+	// 2. Finally, delete the old AzureMachinePool
+	err = r.deleteOldAzureMachinePool(ctx, oldAzureMachinePool)
+	if err != nil {
+		// AzureMachinePool status is updated, so we don't cancel the
+		// reconciliation, other handlers can continue working. It will try to
+		// delete the old AzureMachinePool again in the next reconciliation loop.
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func areReferencesUpdated(azureMachinePool capzexp.AzureMachinePool) bool {
+	// check MachinePool owner reference
+	for _, ref := range azureMachinePool.ObjectMeta.OwnerReferences {
+		if ref.Kind == "MachinePool" && ref.APIVersion == oldcapiexpv1alpha3.GroupVersion.String() {
+			return false
+		}
+	}
+
+	return true
+}
+
+func isStatusEmpty(azureMachinePool capzexp.AzureMachinePool) bool {
+	return reflect.DeepEqual(azureMachinePool.Status, capzexp.AzureMachinePoolStatus{})
+}
+
+func cloneObject(oldObject interface{}, newObject interface{}) error {
```

Comment:
interesting, I expected there would be a builtin function to do this

Reply:
go or CAPI builtin?

There is this `DeepCopy` https://github.com/kubernetes-sigs/cluster-api/blob/main/exp/api/v1alpha3/zz_generated.deepcopy.go#L32, but we want to clone from old MachinePool type to new MachinePool type, and those are different Go types, so we cannot use existing Go func from CAPI.

Reply:
There is go `copy` func, but AFAIU it works with slices.

Reply:
yeah all good, it was just that I thought there was something, but if this works just use this.